docs/hot_module_reloading.md

```diff
@@ -0,0 +1,190 @@
+## What is Hot Module Reloading?
+
+This is a feature that allows you to speed up development process by reducing time between writing
+code and seeing the result. It detects changes in your code and automatically reloads minimum amount
+js modules on the page for changes to take effect.
+
+If something depends on changed module A, it may have some instances of A's types, functions and
+objects closured, so reloading A itself doesnt invalidate closure. In basic configuration
+hot-reloading reloads all modules in chain from changed module up to the root module of applications
+and reruns `main()` function for changes to take effect. This means that without special measures
+state of your won't be preserved, nevertheless hot-reloading is much faster than full page
```

Comment:
[nit] This means that without special measures the state* of your application* won't be preserved